### PR TITLE
Stride order broadcast update

### LINKED ISSUE #1369: codegen performance degradation issue for fusion with stride_order set on inputs
In the example below, enabling stride order doesn't help with indexing, since we are just re-ordering broadcast dimensions. But it's to my surprising that this causes perf regression. Opening this issue for me to track and resolve this.
```
root@913ae90be131:/opt/pytorch/nvfuser# NVFUSER_DUMP=dump_eff_bandwidth python test.py 
kernelinner_persistent_f0_c1_r0_g0 run in 0.035584 ms, achieved: 580.691 GB/s
root@913ae90be131:/opt/pytorch/nvfuser# NVFUSER_DUMP=dump_eff_bandwidth python test.py repro
kernelinner_persistent_f0_c1_r0_g0 run in 0.048832 ms, achieved: 423.151 GB/s
```

Repro scripts below:
```
import torch
from nvfuser import FusionDefinition, DataType

import sys

repro = False
if len(sys.argv) == 2 and sys.argv[1].startswith("repro"):
    repro = True

def nvfuser_fusion_id(fd : FusionDefinition) -> None :
    if repro:
        T0 = fd.define_tensor(shape=[-1, -1, -1], contiguity=[True, True, True], dtype=DataType.BFloat16, is_cpu=False, stride_order=[2, 1, 0])
        T1 = fd.define_tensor(shape=[-1, -1, -1], contiguity=[True, None, None], dtype=DataType.BFloat16, is_cpu=False, stride_order=[1, 0, 2])
        T2 = fd.define_tensor(shape=[-1, -1, -1], contiguity=[True, None, None], dtype=DataType.BFloat16, is_cpu=False, stride_order=[1, 0, 2])
        T3 = fd.define_tensor(shape=[-1, -1, -1], contiguity=[True, True, None], dtype=DataType.BFloat16, is_cpu=False, stride_order=[0, 2, 1])
    else:
        T0 = fd.define_tensor(shape=[-1, -1, -1], contiguity=[True, True, True], dtype=DataType.BFloat16, is_cpu=False)
        T1 = fd.define_tensor(shape=[-1, -1, -1], contiguity=[None, None, True], dtype=DataType.BFloat16, is_cpu=False)
        T2 = fd.define_tensor(shape=[-1, -1, -1], contiguity=[None, None, True], dtype=DataType.BFloat16, is_cpu=False)
        T3 = fd.define_tensor(shape=[-1, -1, -1], contiguity=[None, True, True], dtype=DataType.BFloat16, is_cpu=False)
    T4 = fd.ops.cast(T0, dtype=DataType.Float)
    T5 = fd.ops.cast(T3, dtype=DataType.Float)
    T6 = fd.ops.add(T4, T5)
    S7 = fd.define_scalar(0.00000, dtype=DataType.Double)
    S8 = fd.define_scalar(1.00000, dtype=DataType.Double)
    S9 = fd.define_scalar(16, dtype=DataType.Int)
    S10 = fd.define_scalar(128, dtype=DataType.Int)
    S11 = fd.define_scalar(768, dtype=DataType.Int)
    V12 = fd.define_vector([S9, S10, S11], dtype=DataType.Int)
    T13 = fd.ops.uniform(S7, S8, shape=V12, dtype=DataType.BFloat16)
    S14 = fd.define_scalar(0.900000, dtype=DataType.Double)
    T15 = fd.ops.lt(T13, S14)
    T16 = fd.ops.cast(T15, dtype=DataType.Float)
    T17 = fd.ops.mul(T6, T16)
    S18 = fd.define_scalar(1.11111, dtype=DataType.Double)
    T19 = fd.ops.mul(T17, S18)
    T20, T21 = fd.ops.var_mean(T19, axes=[2], correction=0, keepdim=False)
    S22 = fd.define_scalar(16, dtype=DataType.Int)
    S23 = fd.define_scalar(128, dtype=DataType.Int)
    S24 = fd.define_scalar(1, dtype=DataType.Int)
    V25 = fd.define_vector([S22, S23, S24], dtype=DataType.Int)
    T26 = fd.ops.broadcast_in_dim(T20, shape=V25, broadcast_dims=[0, 1])
    S27 = fd.define_scalar(16, dtype=DataType.Int)
    S28 = fd.define_scalar(128, dtype=DataType.Int)
    S29 = fd.define_scalar(1, dtype=DataType.Int)
    V30 = fd.define_vector([S27, S28, S29], dtype=DataType.Int)
    T31 = fd.ops.broadcast_in_dim(T21, shape=V30, broadcast_dims=[0, 1])
    S32 = fd.define_scalar(1.00000e-05, dtype=DataType.Double)
    T33 = fd.ops.add(T26, S32)
    T34 = fd.ops.rsqrt(T33)
    S35 = fd.define_scalar(16, dtype=DataType.Int)
    S36 = fd.define_scalar(128, dtype=DataType.Int)
    S37 = fd.define_scalar(768, dtype=DataType.Int)
    V38 = fd.define_vector([S35, S36, S37], dtype=DataType.Int)
    T39 = fd.ops.broadcast_in_dim(T31, shape=V38, broadcast_dims=[0, 1, 2])
    T40 = fd.ops.sub(T19, T39)
    S41 = fd.define_scalar(16, dtype=DataType.Int)
    S42 = fd.define_scalar(128, dtype=DataType.Int)
    S43 = fd.define_scalar(768, dtype=DataType.Int)
    V44 = fd.define_vector([S41, S42, S43], dtype=DataType.Int)
    T45 = fd.ops.broadcast_in_dim(T34, shape=V44, broadcast_dims=[0, 1, 2])
    T46 = fd.ops.mul(T40, T45)
    T47 = fd.ops.cast(T1, dtype=DataType.Float)
    T48 = fd.ops.mul(T46, T47)
    T49 = fd.ops.cast(T2, dtype=DataType.Float)
    T50 = fd.ops.add(T48, T49)
    T51 = fd.ops.cast(T50, dtype=DataType.BFloat16)
    fd.add_output(T15)
    fd.add_output(T19)
    fd.add_output(T21)
    fd.add_output(T34)
    fd.add_output(T40)
    fd.add_output(T51)


with FusionDefinition() as fd:
    nvfuser_fusion_id(fd)

inputs = [
    torch.randn((16, 128, 768), dtype=torch.bfloat16, device='cuda:0'),
    torch.randn((768), dtype=torch.bfloat16, device='cuda:0').as_strided((16, 128, 768), (0, 0, 1)),
    torch.randn((768), dtype=torch.bfloat16, device='cuda:0').as_strided((16, 128, 768), (0, 0, 1)),
    torch.randn((128, 768), dtype=torch.bfloat16, device='cuda:0').as_strided((16, 128, 768), (0, 768, 1)),
]
fd.execute(inputs)

```

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixes #1369 

Updated stride_order computation to preserve the semantic position for implicit broadcast dimension, i.e. dimension with stride == 0.
This avoids unnecessary alloc_domain for expanded dimension, which would temporarily avoid perf regression due to the lack of alloc_domain support in many of our schedulers.

This is not a permanent solution. We would still want to plumb through alloc_domain support on schedulers as well as vectorization analysis. Nevertheless, the updated logic is still an improvement.

### CODE REVIEW

**Comment by jjsjann123** [comment]
!build

**Comment by jjsjann123** [comment]
!build

File: python_tests/test_python_frontend.py [hunk 1]
```diff
@@ -1387,25 +1387,61 @@ def test_compute_contiguity(self):
...
-        contiguity = [False, None, True, True, None]
-        stride_order = [4, 2, 3, 0, 1]
-        computed_contiguity, computed_stride_order = compute_tensor_descriptor(
-            sizes, strides
+        configs = (
+            (
+                # size
+                [2, 1, 3, 1, 4, 3],
+                # stride
+                [12, 4, 4, 4, 1, 0],
```
Comment (jacobhinkle) [inline python_tests/test_python_frontend.py:1395]: Should this be [12, 12, 4, 4, 1, 0] to match pytorch?

File: python_tests/test_python_frontend.py [hunk 2]
```diff
@@ -1387,25 +1387,61 @@ def test_compute_contiguity(self):
...
+                # expected contiguity
+                [True, None, True, None, True, None],
+                # expected stride_order
+                [5, 4, 3, 2, 1, 0],
+            ),
+            (
+                [2, 3, 1, 5, 4],
+                [28, 4, 14, 0, 1],
```
Comment (jacobhinkle) [inline python_tests/test_python_frontend.py:1403]: Is the third dim counted as contiguous even though its stride is 14 instead of 12 because the size is one?

File: python_tests/test_python_frontend.py [hunk 3]
```diff
@@ -1387,25 +1387,61 @@ def test_compute_contiguity(self):
...
+                [False, None, True, None, True],
+                [4, 2, 3, 1, 0],
+            ),
+            (
+                [2, 2, 1, 1, 2, 2, 2],
+                [8, 4, 3, 9, 2, 0, 1],
```
Comment (jacobhinkle) [inline python_tests/test_python_frontend.py:1409]: Are aliasing strides allowed?

**Review by jacobhinkle — COMMENTED** [review]
Just had a couple questions on test cases.

File: python_tests/test_python_frontend.py (hunk 3 above)
Comment (jjsjann123) [reply python_tests/test_python_frontend.py:1409]: There's no memory overlapping here. I think that's what you meant by aliasing strides.
i.e. the *funny* strides are all in dimension with size-1. So we aren't really indexing or using those at all.

File: python_tests/test_python_frontend.py (hunk 2 above)
Comment (jjsjann123) [reply python_tests/test_python_frontend.py:1403]: I think there's a confusion here.

The `expected contiguity` flag is marked with the sorted stride. That's also how we set contiguity flag in c++ API, regarding alloc_domain.

So our alloc_domain is permuted regarding the stride_order. i.e.
sorted size:    [2, 1, 3, 5, 4]
sorted stride:  [28, 14, 4, 0, 1]
expected contiguity flag: [ False, None, True, None, True]

broadcast dimension with size 1 and stride 14 is marked as `None`, since it's a broadcast here.

File: python_tests/test_python_frontend.py (hunk 1 above)
Comment (jjsjann123) [reply python_tests/test_python_frontend.py:1395]: I'm not making assumption on how the tensor is constructed. I can add the size and stride if you think it's of interest. :wink:

File: python_tests/test_python_frontend.py (hunk 2 above)
Comment (jacobhinkle) [reply python_tests/test_python_frontend.py:1403]: Oh of course. My mistake. Thanks for the explanation.

File: python_tests/test_python_frontend.py (hunk 3 above)
Comment (jacobhinkle) [reply python_tests/test_python_frontend.py:1409]: Right. Got it. Thanks.

File: python_tests/test_python_frontend.py (hunk 1 above)
Comment (jacobhinkle) [reply python_tests/test_python_frontend.py:1395]: Not necessary. Just making sure I understand the intention.

Comment (jacobhinkle) [reply python_tests/test_python_frontend.py:1395]: Not necessary. Just making sure I understand the intention.

File: csrc/python_frontend/python_bindings.cpp [hunk 4]
```diff
@@ -271,29 +271,43 @@ std::vector<std::optional<bool>> computeContiguity(
...
-//   contiguity would be: [False, None, True, True, None]
+//   sizes   = [2, 1, 3, 1, 4]
+//   strides = [24, 4, 8, 4, 2]
+// Note that there are a few explicit broadcast dimensions, dimensions with size
+// == 1. The stride for explicit broadcast dimensions participates in stride
+// order computation. The reason is that, frameworks could assign meaningful
+// stride to an explicit broadcast dimensions to hint memory format, which could
+// be used to deduce the desired output memory format. We use stable sort to
+// break tie when two dimension has equal stride, i.e. try to preserve their
+// semantical order. Hence, we would compute stride order as: [4, 2, 3, 1, 0].
```
Comment (wujingyue) [inline csrc/python_frontend/python_bindings.cpp:294]: > semantical

Is there a better term for this? Maybe "relative input order"? Feel free to ignore.

File: csrc/python_frontend/python_bindings.cpp [hunk 5]
```diff
@@ -302,19 +316,33 @@ computeTensorDescriptor(
...
-      "compute_tensor_descriptor: Sizes and strides must have the same number of dimensions");
+      "compute_tensor_descriptor: "
+      "Sizes and strides must have the same number of dimensions");
   std::vector<DimInfo> dim_info_vec;
+  std::vector<DimInfo> broadcast_dim_vec;
   for (auto i : c10::irange(sizes.size())) {
-    // NOTE: not supporting negative stride yet.
+    // NOTE: not supporting negative stride yet, but we can probably allow it on
+    // broadcast dims
     NVF_CHECK(strides[i] >= 0, "negative stride on tensor is not supported");
```
Comment (wujingyue) [inline csrc/python_frontend/python_bindings.cpp:326]:
```suggestion
    NVF_CHECK(strides[i] >= 0, "negative stride on tensor is not supported: strides[", i, "]=", strides[i]);
```

File: csrc/python_frontend/python_bindings.cpp [hunk 6]
```diff
@@ -302,19 +316,33 @@ computeTensorDescriptor(
...
-    dim_info_vec.emplace_back(DimInfo{(int64_t)i, sizes[i], strides[i]});
+    DimInfo dim_info = DimInfo{(int64_t)i, sizes[i], strides[i]};
```
Comment (wujingyue) [inline csrc/python_frontend/python_bindings.cpp:327]:
```suggestion
    DimInfo dim_info{(int64_t)i, sizes[i], strides[i]};
```

File: csrc/python_frontend/python_bindings.cpp [hunk 7]
```diff
@@ -302,19 +316,33 @@ computeTensorDescriptor(
     const std::vector<int64_t>& strides) {
   NVF_CHECK(
       sizes.size() == strides.size(),
-      "compute_tensor_descriptor: Sizes and strides must have the same number of dimensions");
+      "compute_tensor_descriptor: "
+      "Sizes and strides must have the same number of dimensions");
   std::vector<DimInfo> dim_info_vec;
```
Comment (wujingyue) [inline csrc/python_frontend/python_bindings.cpp:321]:
```suggestion
  std::vector<DimInfo> non_broadcast_dim_info_vec;
```

File: csrc/python_frontend/python_bindings.cpp [hunk 8]
```diff
@@ -302,19 +316,33 @@ computeTensorDescriptor(
...
+  std::vector<DimInfo> broadcast_dim_vec;
```
Comment (wujingyue) [inline csrc/python_frontend/python_bindings.cpp:322]:
```suggestion
  // Collects implicit broadcast dimensions.
  std::vector<DimInfo> broadcast_dim_vec;
```

File: csrc/python_frontend/python_bindings.cpp [hunk 9]
```diff
@@ -271,29 +271,43 @@ std::vector<std::optional<bool>> computeContiguity(
...
-// stride_order would be: [4, 2, 3, 0, 1], marking the order of strides in the
-// vector. Meanwhile, contiguity would be computed on the physical domain, i.e.
-// on sorted sizes & strides.
-//   sorted_size    = [2, 1, 3, 4, 5]
-//   sorted_strides = [28, 14, 4, 1, 0]
-//   contiguity would be: [False, None, True, True, None]
+//   sizes   = [2, 1, 3, 1, 4]
+//   strides = [24, 4, 8, 4, 2]
+// Note that there are a few explicit broadcast dimensions, dimensions with size
+// == 1. The stride for explicit broadcast dimensions participates in stride
```
Comment (wujingyue) [inline csrc/python_frontend/python_bindings.cpp:289]: Is <size=1, stride=0> considered implicit or explicit?

**Review by wujingyue — APPROVED** [review]

File: csrc/python_frontend/python_bindings.cpp (hunk 9 above)
Comment (jjsjann123) [reply csrc/python_frontend/python_bindings.cpp:289]: Good catch!

The documentation here says it's explicit, but the implementation is treating it as implicit, since it only checks for `stride == 0`.
Let me clean it up and add a test for that.

**Comment by jjsjann123** [comment]
!build
